test/Microsoft.Azure.Devices.Edge.Test.Common/LeafDevice.cs

```diff
@@ -0,0 +1,352 @@
+        {
+            try
+            {
+                return await what();
+            }
+            catch
+            {
+                await DeleteIdentityAsync(device, iotHub, token);
+                throw;
+            }
+        }
+
+        static async Task<LeafDevice> CreateLeafDeviceAsync(Device device, Func<DeviceClient> clientFactory, IotHub iotHub, CancellationToken token)
+        {
+            DeviceClient client = clientFactory();
+            await client.SetMethodHandlerAsync("DirectMethod", DirectMethod, null, token);
+            return new LeafDevice(device, client, iotHub);
+        }
+
+        public Task SendEventAsync(CancellationToken token)
+        {
+            var message = new Message(Encoding.ASCII.GetBytes(this.device.Id))
+            {
+                Properties = { ["leaf-message-id"] = this.messageId }
+            };
+            return this.client.SendEventAsync(message, token);
+        }
+
+        public Task WaitForEventsReceivedAsync(DateTime seekTime, CancellationToken token)
+        {
+            return Profiler.Run(
+                () => this.iotHub.ReceiveEventsAsync(
+                    this.device.Id,
+                    seekTime,
+                    data =>
+                    {
+                        data.SystemProperties.TryGetValue("iothub-connection-device-id", out object devId);
+                        data.Properties.TryGetValue("leaf-message-id", out object msgId);
+
+                        return devId != null && devId.ToString().Equals(this.device.Id)
+                                             && msgId != null && msgId.ToString().Equals(this.messageId);
+                    },
+                    token),
+                "Received events from device '{Device}' on Event Hub '{EventHub}'",
+                this.device.Id,
+                this.iotHub.EntityPath);
+        }
+
+        public Task InvokeDirectMethodAsync(CancellationToken token) =>
+            Profiler.Run(
+                () => this.iotHub.InvokeMethodAsync(
+                    this.device.Id,
+                    new CloudToDeviceMethod("DirectMethod"),
+                    token),
+                "Invoked method on leaf device from the cloud");
+
+        // BUG: callers can continue to try to use this object after DeleteIdentityAsync has been called.
+        public Task DeleteIdentityAsync(CancellationToken token) =>
+            DeleteIdentityAsync(this.device, this.iotHub, token);
+
+        static Task DeleteIdentityAsync(Device device, IotHub iotHub, CancellationToken token) =>
+            Profiler.Run(
+                () => iotHub.DeleteDeviceIdentityAsync(device, token),
+                "Deleted device '{Device}'",
+                device.Id);
+
+        static Task<MethodResponse> DirectMethod(MethodRequest request, object context)
```

Comment:
```suggestion
        static Task<MethodResponse> DirectMethodCallbackHandler(MethodRequest request, object context)
```

Reply:
Then we'd have to register the direct method name "DirectMethodCallbackHandler" with the SDK as well, which doesn't make sense. Either that, or add a const string for "DirectMethod" and pass that to `CloudToDeviceMethod` and `SetMethodHandlerAsync` instead of using `nameof()`. The proposed name isn't a clear winner over the current name, so I vote we leave it alone.